### PR TITLE
MAINT/CI: special/array types: test alternative backends in CI

### LINKED ISSUE #19928: TST: special: array types: test tol failure with `torch` backend
cc @mdhaber 
```
_________________________________________________________________________ test_support_alternative_backends[f_name_n_args9-torch] __________________________________________________________________________
scipy/special/tests/test_support_alternative_backends.py:24: in test_support_alternative_backends
    @given(data=strategies.data())
        f          = <function given.<locals>.run_test_as_given.<locals>.wrapped_test at 0x123cbe9e0>
        f_name_n_args = ('gammaln', 1)
        xp         = <module 'torch' from '/Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/torch/__init__.py'>
scipy/special/tests/test_support_alternative_backends.py:51: in test_support_alternative_backends
    xp_assert_close(res, xp.asarray(ref, dtype=dtype_xp), rtol=eps**0.5, atol=eps*10,
        args_np    = [array(-10.)]
        args_xp    = [tensor(-10.0000, dtype=torch.float64)]
        data       = data(...)
        dtype      = 'float64'
        dtype_np   = <class 'numpy.float64'>
        dtype_xp   = torch.float64
        elements   = {'allow_subnormal': False, 'max_value': 10.0, 'min_value': -10.0}
        eps        = 2.220446049250313e-16
        f          = <function support_alternative_backends.<locals>.wrapped at 0x11f548c10>
        f_name     = 'gammaln'
        f_name_n_args = ('gammaln', 1)
        final_shape = ()
        mbs        = mutually_broadcastable_shapes(num_shapes=1)
        n_args     = 1
        ref        = array(5.80645377)
        res        = tensor(5.8065, dtype=torch.float64)
        shapes     = ((),)
        xp         = <module 'torch' from '/Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/torch/__init__.py'>
scipy/_lib/_array_api.py:293: in xp_assert_close
    return xp.testing.assert_close(actual, desired, rtol=rtol, atol=atol,
E   AssertionError: Scalars are not close!
E   
E   Absolute difference: 1.0175748776930504e-07 (up to 2.220446049250313e-15 allowed)
E   Relative difference: 1.7524894158750067e-08 (up to 1.4901161193847656e-08 allowed)
E   Falsifying example: test_support_alternative_backends(
E       xp=<module 'torch' from '/Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/torch/__init__.py'>,
E       f_name_n_args=('gammaln', 1),
E       data=data(...),
E   )
E   Draw 1: BroadcastableShapes(input_shapes=((),), result_shape=())
E   Draw 2: 'float64'
E   Draw 3: array(-10.)
E   Explanation:
E       These lines were always and only run by failing examples:
E           /Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/torch/testing/_comparison.py:1000
E           /Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/torch/testing/_comparison.py:1511
E           /Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/torch/testing/_comparison.py:174
E           /Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/typing.py:308
E   
E   You can reproduce this example by temporarily adding @reproduce_failure('6.92.2', b'AAAAAAEAAAABgDMJDj////8=') as a decorator on your test case
        actual     = tensor(5.8065, dtype=torch.float64)
        atol       = 2.220446049250313e-15
        check_dtype = True
        check_namespace = True
        check_shape = True
        desired    = tensor(5.8065, dtype=torch.float64)
        err_msg    = None
        rtol       = 1.4901161193847656e-08
        xp         = <module 'scipy._lib.array_api_compat.torch' from '/Users/lucascolley/programming/scipy/build-install/lib/python3.10/site-packages/scipy/_lib/array_api_compat/torch/__init__.py'>
```

### PR DESCRIPTION
**Pull request by lucascolley**
#### Reference issue
Closes gh-19928

#### What does this implement/fix?
Getting the tests to pass both locally for me and in CI

#### Additional information

cc @mdhaber

### CODE REVIEW

**Comment by mdhaber** [comment]
Did it need to be *that* loose? That's like 3e-2 for double precision.

**Comment by lucascolley** [comment]
> Did it need to be that loose?

`eps**0.25` failed. I can try something in-between if you'd like.

**Comment by mdhaber** [comment]
The failure in gh-19928 was:
```
E   Absolute difference: 1.0175748776930504e-07 (up to 2.220446049250313e-15 allowed)
E   Relative difference: 1.7524894158750067e-08 (up to 1.4901161193847656e-08 allowed)
```
suggesting that only a tiny bump would be required.

Could you try merging main locally? Maybe this needs gh-19931?

**Comment by lucascolley** [comment]
> Could you try merging main locally? Maybe this needs https://github.com/scipy/scipy/pull/19931?

I rebased on top of `main` and I still see:

```
E   AssertionError: Scalars are not close!
E   
E   Absolute difference: 0.054916016679925406 (up to 2.220446049250313e-15 allowed)
E   Relative difference: 0.0016505638526979993 (up to 0.0001220703125 allowed)
E   Falsifying example: test_support_alternative_backends(
E       xp=<module 'torch' from '/Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/torch/__init__.py'>,
E       f_name_n_args=('gammaln', 1),
E       data=data(...),
E   )
E   Draw 1: BroadcastableShapes(input_shapes=((),), result_shape=())
E   Draw 2: 'float64'
E   Draw 3: array(-2.)
E   Explanation:
E       These lines were always and only run by failing examples:
E           /Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/torch/testing/_comparison.py:1000
E           /Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/torch/testing/_comparison.py:1511
E           /Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/torch/testing/_comparison.py:174
E           /Users/lucascolley/mambaforge/envs/scipy-dev/lib/python3.10/typing.py:308
E   
E   You can reproduce this example by temporarily adding @reproduce_failure('6.92.2', b'AAAAAAEAAAABgAH///////8=') as a decorator on your test case
```

**Comment by mdhaber** [comment]
Are you able to isolate the input that is causing the problem and try it manually? That seems like a really large difference that we should explain.

**Comment by lucascolley** [comment]
```
(Pdb) args_np
[array(-2.)]
(Pdb) args_xp
[tensor(-2.0000, dtype=torch.float64)]
(Pdb) from scipy.special import gammaln
(Pdb) gammaln(*args_np)
33.27106466687738
(Pdb) gammaln(*args_xp)
tensor(33.2161, dtype=torch.float64)
```

**Comment by steppi** [comment]
The actual value must be slightly different from `-2.0` since that's a pole of the gamma function. I did a quick check and it it's

```python
import scipy.special as special
from mpmath import mp

special.gammaln(-1.9999999999999982)
# 33.27106466687738

mp.loggamma(-1.9999999999999982)
# mpc(real='33.271064666877379', imag='-6.2831853071795862')
```

Note that `mpmath` has chosen a different branch, but otherwise agrees very closely with SciPy. PyTorch seems to struggle near the poles.

**Comment by mdhaber** [comment]
Hmmm well I guess we can't blame Hypothesis for exploiting a weak spot. Can you think of a simple way to exclude the poles? I set up those tests in the simplest way that seemed to pass, without being too careful about the domains each functions separately. If worst comes to worst, maybe we generate the evalutation points with a Generator instead of Hypothesis.

**Comment by steppi** [comment]
> Hmmm well I guess we can't blame Hypothesis for exploiting a weak spot. Can you think of a simple way to exclude the poles? I set up those tests in the simplest way that seemed to pass, without being too careful about the domains each functions separately. If worst comes to worst, maybe we generate the evalutation points with a Generator instead of Hypothesis.

Not sure if there’s an easy way off hand. Since we just want to test that they’re supposed to compute the same function, maybe it’s fine to loosen the tolerances as drastically as @lucascolley suggested but also check that the root mean square error is less then `eps**0.5`?

**Comment by mdhaber** [comment]
> but also check that the root mean square error is less then eps**0.5?

OK, sure.

File: .github/workflows/array_api.yml [hunk 1]
```diff
@@ -91,7 +91,7 @@ jobs:
...
         export SCIPY_USE_PROPACK=1
         # expand as new modules are added
-        # --array-api-backend == -b
-        python dev.py --no-build test -b pytorch -b numpy -b array_api_strict -s cluster -- --durations 3 --timeout=60
-        python dev.py --no-build test -b pytorch -b numpy -b array_api_strict -s fft -- --durations 3 --timeout=60
-        python dev.py --no-build test --array-api-backend all --tests scipy/_lib/tests/test_array_api.py
+        python dev.py --no-build test -b all -s cluster -- --durations 3 --timeout=60
+        python dev.py --no-build test -b all -s fft -- --durations 3 --timeout=60
+        python dev.py --no-build test -b all -t scipy.special.tests.test_support_alternative_backends -- --durations 3 --timeout=60
+        python dev.py --no-build test -b all -t scipy._lib.tests.test_array_api.py
```
Comment (lucascolley) [inline .github/workflows/array_api.yml:97]:
```suggestion
        python dev.py --no-build test -b all -t scipy._lib.tests.test_array_api
```

File: scipy/special/tests/test_support_alternative_backends.py [hunk 2]
```diff
@@ -55,5 +55,13 @@ def test_support_alternative_backends(xp, data, f_name_n_args):
     res = f(*args_xp)
 
     eps = np.finfo(dtype).eps
-    xp_assert_close(res, xp.asarray(ref, dtype=dtype_xp), rtol=eps**0.5, atol=eps*10,
-                    check_namespace=True, check_shape=True, check_dtype=True)
+    # PyTorch seems to struggle with precision near the poles of `gammaln`,
+    # so the tolerance needs to be quite loose (eps**0.2) - see gh-19935.
+    # To compensate, we aso check that that the root-mean-square error is
```
Comment (lucascolley) [inline scipy/special/tests/test_support_alternative_backends.py:60]:
```suggestion
    # To compensate, we also check that that the root-mean-square error is
```

**Comment by tylerjereddy** [comment]
Just a drive by comment, but if you want to avoid the <= 0 integer poles of the gamma function, you could also tell `hypothesis` to so with https://hypothesis.readthedocs.io/en/latest/details.html#making-assumptions. That's what I normally do. I guess you're testing more than just that special function here so you might need something like (pseudo-code):

```python
if func == "gamma_func":
    assume not (isinstance(x, int) and x <= 0)
```

Not sure on the tradeoffs vs. what you're doing at the moment. I believe one problem with my suggestion is that if other special functions have different weaknesses/poles/whatever, then the `hypothesis` "algorithm" may avoid these values as a result of the `assume` call getting hit a few times. That might be a case for splitting off to separate tests at some point I suppose, depending on special function weaknesses. Maybe that's a bit too far for now, but wanted to mention.

**Comment by steppi** [comment]
@tylerjereddy. That could work, modulo it’s not actually the poles that are the problem in PyTorch but points very close to the poles. For now, these tests are just intended for checking that these are actually supposed to be equivalent functions, which can be an issue for e.g. alternative parameterizations. Individual tests for each function make a lot of sense, but I think should be a separate thing, and maybe shouldn’t even live in SciPy.

**Comment by steppi** [comment]
@lucascolley, if you appease the linter by removing the offending Unicode character and the CI settles, I think this will be ready to merge.

File: scipy/special/tests/test_support_alternative_backends.py [hunk 3]
```diff
@@ -55,5 +55,13 @@ def test_support_alternative_backends(xp, data, f_name_n_args):
     res = f(*args_xp)
 
     eps = np.finfo(dtype).eps
-    xp_assert_close(res, xp.asarray(ref, dtype=dtype_xp), rtol=eps**0.5, atol=eps*10,
-                    check_namespace=True, check_shape=True, check_dtype=True)
+    # PyTorch seems to struggle with precision near the poles of `gammaln`,
+    # so the tolerance needs to be quite loose (eps**0.2) - see gh-19935.
+    # To compensate, we also check that that the root-mean-square error is
```
Comment (lucascolley) [inline scipy/special/tests/test_support_alternative_backends.py:60]:
```suggestion
    # To compensate, we also check that the root-mean-square error is
```

**Comment by lucascolley** [comment]
oops - that'll teach me to avoid doing git surgery in codespaces

**Pull request closed without merging**